### PR TITLE
Creating ScrollBar as a seperate UI element

### PR DESCRIPTION
**Pull request by ganimtron-10**
By Rferencing PR #285 
Implementing Scrollbar as an individual element.
This makes ScrollBar more flexible and code duplicacy is avoided.

scrollbar

### CODE REVIEW

**Review by m-agour — COMMENTED** [review]
Hello @ganimtron-10 ,
Good job!
I only have one minor note: The scroll bar only changes color when being scrolled. It would be better to change its color on click or on hover.
Thanks.

**Review by xtanion — COMMENTED** [review]
Hey @ganimtron-10 , Everything looks fine. The scroll handle shows the Red color `(0.9, 0.2, 0.2)` initially, and then it switches to `active_color` or `inactive_color`.

File: fury/ui/elements.py [hunk 1]
```diff
@@ -2294,6 +2295,193 @@ def left_button_dragged(self, i_ren, _obj, _sub_component):
...
+        """
+        Parameters
+        ----------
+        size : (int, int)
+            Size of the scrollbar in height & width.
+        content_span : int
+            Span of the entire content.
+        view_span : int
+            Span of the viewable area to display the content.
+        track_color : tuple of 3 floats
```
Comment (skoudoro) [inline fury/ui/elements.py:2319]: `, optional` to add

File: fury/ui/elements.py [hunk 2]
```diff
@@ -2294,6 +2295,193 @@ def left_button_dragged(self, i_ren, _obj, _sub_component):
...
+            Color of the scrollbar track.
+        reverse_scrolling : {True, False}
```
Comment (skoudoro) [inline fury/ui/elements.py:2321]: `bool, optional` instead

File: fury/ui/elements.py [hunk 3]
```diff
@@ -2294,6 +2295,193 @@ def left_button_dragged(self, i_ren, _obj, _sub_component):
...
+            If True, scrolling up will move the list of files down.
+        active_color : tuple of 3 floats
```
Comment (skoudoro) [inline fury/ui/elements.py:2323]: `, optional` to add

File: fury/ui/elements.py [hunk 4]
```diff
@@ -2294,6 +2295,193 @@ def left_button_dragged(self, i_ren, _obj, _sub_component):
...
+            Color of the scrollbar when in use.
+        inactive_color : tuple of 3 floats
```
Comment (skoudoro) [inline fury/ui/elements.py:2325]: `, optional` to add

File: fury/ui/elements.py [hunk 5]
```diff
@@ -2294,6 +2295,193 @@ def left_button_dragged(self, i_ren, _obj, _sub_component):
...
+            Color of the scrollbar when not in use.
+        orientation : {"vertical", "horizontal"}
```
Comment (skoudoro) [inline fury/ui/elements.py:2327]: `str, optional` instead

File: fury/ui/elements.py [hunk 6]
```diff
@@ -2294,6 +2295,193 @@ def left_button_dragged(self, i_ren, _obj, _sub_component):
...
+            The orientation of the scrollbar.
+        opacity : float
```
Comment (skoudoro) [inline fury/ui/elements.py:2329]: `, optional` to add

File: fury/ui/elements.py [hunk 7]
```diff
@@ -2294,6 +2295,193 @@ def left_button_dragged(self, i_ren, _obj, _sub_component):
...
+            Opacity of scrollbar.
+        position : (int, int, int)
```
Comment (skoudoro) [inline fury/ui/elements.py:2331]: `, optional` to add

why 3 `int`, there is 2 `int` above

File: fury/ui/elements.py [hunk 8]
```diff
@@ -2294,6 +2295,193 @@ def left_button_dragged(self, i_ren, _obj, _sub_component):
...
+            y_position = min(
+                y_position,
+                self.track.position[1] +
+                self.track.size[1] - self.bar.size[1] // 2,
+            )
+            self.bar.center = (self.track.center[0], y_position)
+
+    def scrolling_callback(self, i_ren, _obj, _rect_obj):
+        """Callback to change the color of the bar when it is clicked.
+        Parameters
```
Comment (skoudoro) [inline fury/ui/elements.py:2388]: need empty line above

File: fury/ui/elements.py [hunk 9]
```diff
@@ -2294,6 +2295,193 @@ def left_button_dragged(self, i_ren, _obj, _sub_component):
...
+        self.bar.color = self.active_color
+        position = i_ren.event.position
+        self.set_position(position)
+        self.on_scroll(self)
+        i_ren.force_render()
+        i_ren.event.abort()
+
+    def release_callback(self, i_ren, _obj, _rect_obj):
+        """Callback to change the color of the bar when it is released.
+        Parameters
```
Comment (skoudoro) [inline fury/ui/elements.py:2404]: need empty line above

**Review by skoudoro — COMMENTED** [review]
Hi @ganimtron-10,

Thanks! 

Can you show how you plan to integrate this new scrollbar UI with ListBox2D or FileMenu2D? 

Also, to give credit to the original authors (@Nibba2018 and @devmodi154), It would be great to restart from his branch, rebase it and add your change after that. 

Last question: What is the difference between the scrollbar ui and the slider UI? 
you can see the slider UI here: https://fury.gl/latest/auto_tutorials/02_ui/viz_ui_slider.html#sphx-glr-auto-tutorials-02-ui-viz-ui-slider-py

I suppose some code can be shared between both elements.

**Comment by ganimtron-10** [comment]
Hello @skoudoro ,


> Can you show how you plan to integrate this new scrollbar UI with ListBox2D or FileMenu2D?

Yeah sure!
Before continuing further I wanted to confirm the approach.



> Also, to give credit to the original authors (@Nibba2018 and @devmodi154), It would be great to restart from his branch, rebase it and add your change after that.

I tried rebasing the code, but because the UI's structure has now changed, for each commit in the old branch it is raising a conflict which has to be solved manually.
I also tried merging the master branch directly but there the original `elements.py` has got modified to old `ui.py` as you can see [here](https://github.com/ganimtron-10/fury/pull/5).
Can you suggest some ways by which I could restart from the old branch??

>Last question: What is the difference between the scrollbar ui and the slider UI?
you can see the slider UI here: https://fury.gl/latest/auto_tutorials/02_ui/viz_ui_slider.html#sphx-glr-auto-tutorials-02-ui-viz-ui-slider-py
I suppose some code can be shared between both elements.

I had a thought about inheriting the slider ui, but then as they have implemented the whole new class from scratch in #285 I continued their work.

**Comment by ganimtron-10** [comment]
Closing with reference to #816.

**Pull request closed without merging**